src/test/java/org/mskcc/cbio/oncokb/web/rest/UsageAnalysisControllerIT.java

```diff
@@ -0,0 +1,957 @@
+  }
+
+  private static class MockS3Data {
+
+    private static class KeyValuePair<Key, Value> {
+      public final Key key;
+      public final Value value;
+
+      public KeyValuePair(Key key, Value value) {
+        this.key = key;
+        this.value = value;
+      }
+    }
+
+    private final String[] sampleOncokbEndpoints = {
+      "/api/v1/annotate/mutations/byProteinChange",
+      "/api/private/utils/numbers/main/",
+      "/api/v1/annotate/structuralVariants",
+      "/api/private/utils/numbers/levels/",
+      "/api/v1/annotate/copyNumberAlterations",
+      "/api/private/search/typeahead",
+    };
+    private final String[] fakeCompanies = { "FakeCo", "DemoLLC" };
+    private final Random random;
+
+    private final JsonObject files = new JsonObject();
+
+    private final List<User> users = new ArrayList<>();
+    private final List<UserDTO> userDtos = new ArrayList<>();
+
+    public MockS3Data(int seed) throws Exception {
+      this(seed, new String[] { "doej", "smithj" });
+    }
+
+    public MockS3Data(int seed, String[] userNames) throws Exception {
+      random = new Random(seed);
+      addUsers(userNames);
+    }
+
+    private String getNextOncokbEndpoint() {
+      return sampleOncokbEndpoints[Math.abs(
+          random.nextInt() % sampleOncokbEndpoints.length
+        )];
+    }
+
+    /**
+     * Get value from files JSON object based on the JSON path created by the
+     * {@code keys}. If an object is missing in the JSON path then it's created.
+     * @param firstKey firstKey in JSON path
+     * @param restOfTheKeys The keys in JSON path
+     * @return The resulting {@code KeyValuePair}
+     */
+    private KeyValuePair<String, JsonObject> safeFetchJsonObjectFromFilesObject(
+      String firstKey,
+      String... restOfTheKeys
+    ) {
+      JsonObject obj = files;
+      if (!obj.has(firstKey)) {
+        obj.add(firstKey, new JsonObject());
+      }
+      obj = (JsonObject) obj.get(firstKey);
+
+      for (int i = 0; i < restOfTheKeys.length - 1; i++) {
+        String key = restOfTheKeys[i];
+        if (!obj.has(key)) {
+          obj.add(key, new JsonObject());
+        }
+        obj = (JsonObject) obj.get(key);
+      }
+      String lastKey = restOfTheKeys.length == 0
+        ? firstKey
+        : restOfTheKeys[restOfTheKeys.length - 1];
+      return new KeyValuePair<>(lastKey, obj);
+    }
+
+    /**
+     * Get value from files JSON object based on the JSON path created by the
+     * {@code firstKey}/[{@code index}]/{@code keys}.
+     * If an object is missing in the JSON path then it's created.
+     * @param firstKey First key in the files object.
+     * @param index Array index of the array the {@code firstKey} parameter points to.
+     * @param restOfTheKeys The remaining JSON path keys.
+     * @return The resulting {@code KeyValuePair}
+     */
+    private KeyValuePair<String, JsonObject> safeFetchJsonObjectFromFilesObject(
+      String firstKey,
+      int index,
```

Comment:
Kind of. It's assuming the first key points to an array of JSON objects.

Yes, it is assumed that the rest of the values are JSON objects.

It is risky to assume there will be an array, but it's only being used for this test code. The class is also scoped to just this test class so there isn't a risk someone will use it outside of this class unless they change the code. I don't it's worth making this code more generic if this is only place it's used.

If we want absolute type safety then I should switch to classes instead of building out JSON.